upstairs/src/deferred.rs

```diff
@@ -239,13 +239,24 @@ pub(crate) struct DeferredMessage {
     pub hashes: Vec<Option<u64>>,
 
     pub client_id: ClientId,
+
+    /// See `DeferredRead::connection_id`
+    pub connection_id: u64,
 }
 
 /// Standalone data structure which can perform decryption
 pub(crate) struct DeferredRead {
     /// Message, which must be a `ReadResponse`
     pub message: Message,
 
+    /// Unique ID for this particular connection to the downstairs
+    ///
+    /// This is needed because -- if read decryption is deferred -- it may
+    /// complete after we have disconnected from the client, which would make
+    /// handling the decrypted value incorrect (because it may have been skipped
+    /// or re-sent).
+    pub connection_id: u64,
```

Comment:
Just to be sure I understand it, we are really checking to be sure that the connection ID
at the time we took a message off the wire from a downstairs matches the connection ID
when we process the job in `on_client_message()`?

Like, we don't need to check for, and don't expect to see, a difference in connection ID
when we put a message on the wire and when we pull it off, right?
Any "in flight" messages won't make it back if we stop the client task.

Reply:
> Like, we don't need to check for, and don't expect to see, a difference in connection ID
> when we put a message on the wire and when we pull it off, right?
> Any "in flight" messages won't make it back if we stop the client task.

Correct, that was what #1126 fixes.  This adds the same "ignore messages from disconnected downstairs" for messages that were deferred _before_ the downstairs was disconnected, but complete their deferred operation afterwards.